### PR TITLE
lisp_fn exported functions should be public within rust too.

### PR DESCRIPTION
**Pull request by shaleh**
Enforce this by in the build process.
Also, remove export files if there are no longer exported functions.
This prevents odd compilation issues when the sole exported function
in a module has its #[lisp_fn] macro commented out to help debug a
compilation error.

### CODE REVIEW

File: rust_src/build.rs [hunk 1]
```diff
@@ -110,8 +110,19 @@ where
             },
 
             ParseState::LispFn(name) => {
-                if line.starts_with("pub") || line.starts_with("fn") {
-                    if let Some(func) = name.or_else(|| get_function_name(&line)) {
+                let name = name.or_else(|| get_function_name(&line));
+
+                if line.starts_with("fn") {
+                    panic!(format!(
```
Comment (birkenfeld) [inline rust_src/build.rs]: You don't need `format!` in `panic!`, it supports format args out of the box.

File: rust_src/build.rs [hunk 2]
```diff
@@ -110,8 +110,19 @@ where
...
+                        "
+{} is not public in {} module.
```
Comment (birkenfeld) [inline rust_src/build.rs]: ... in *the* ... module?

File: rust_src/build.rs [hunk 3]
```diff
@@ -110,8 +110,19 @@ where
...
+lisp_fn functions are meant to be used from within remacs as well as in lisp code.
+",
+                        name.unwrap_or("<unknown>".to_string()),
```
Comment (birkenfeld) [inline rust_src/build.rs]: If `name` is still `None` (can that even happen?) the error isn't very helpful. Do we have a line number?

File: rust_src/build.rs (hunk 1 above)
Comment (shaleh) [reply rust_src/build.rs]: Which one is it that doesn't? I know one of the macros complains when I give it format args. I thought it was panic.

File: rust_src/build.rs (hunk 3 above)
Comment (shaleh) [reply rust_src/build.rs]: At the moment no. I could try and count.

Comment (shaleh) [reply rust_src/build.rs]: It is highly unlikely that the None case happens.

File: rust_src/build.rs (hunk 1 above)
Comment (birkenfeld) [reply rust_src/build.rs]: I seem to remember something similar, but right now I don't find any. It might have been fixed some Rust versions ago.

Comment (shaleh) [reply rust_src/build.rs]: Good to know.

**Comment by shaleh** [comment]
I am a lot happier with this commit. The error when a failure occurs is much easier to read.

File: rust_src/build.rs [hunk 4]
```diff
@@ -31,14 +32,11 @@ fn get_function_name(line: &str) -> Option<String> {
         }
     }
 
-    None
+    panic!("Failed to find function name in: '{}'", line)
```
Comment (birkenfeld) [inline rust_src/build.rs]: function can now return `String` directly?

File: rust_src/build.rs [hunk 5]
```diff
@@ -110,19 +110,29 @@ where
...
+                if line.starts_with("fn") {
+                    eprintln!(
+                        "
+`{}` is not public in the {} module at line {}.
+lisp_fn functions are meant to be used from Rust as well as in lisp code.",
+                        name,
+                        modname,
+                        lineno,
+                    );
+                    process::exit(1);
```
Comment (birkenfeld) [inline rust_src/build.rs]: Why use panic in some cases and process::exit in others?

Comment (shaleh) [reply rust_src/build.rs]: The panics are exceptions. The exits are user errors.

File: rust_src/build.rs (hunk 4 above)
Comment (shaleh) [reply rust_src/build.rs]: Good point. Although I meant move the panic into the main line.

Comment (shaleh) [reply rust_src/build.rs]: Bah phones. Might not meant.

**Comment by shaleh** [comment]
You can see what a build error looks like. Apparently my branch was not in sync.

**Comment by shaleh** [comment]
toggling build

**Pull request merged**